### PR TITLE
Add CUDA APIs: PeekAtLastError, int_as_float, Atomics

### PR DESCRIPTION
**Pull request by mxmlnkn**
### Implemented CUDA APIs
  - add [`cudaPeekAtLastError`](https://docs.nvidia.com/cuda/cuda-c-programming-guide/#error-checking)
  - add several atomics:
    - atomicSub
    - atomicMin
    - atomicMax
    - atomicInc
    - atomicDec
    - atomicExch
    - atomicCAS
  - add [`__float_as_int`](https://docs.nvidia.com/cuda/cuda-math-api/group__CUDA__MATH__INTRINSIC__CAST.html#group__CUDA__MATH__INTRINSIC__CAST_1g781a84bf3b7efe84969c2dc9b0915fa1) function

### Documentation Change
  - proper code indentation inside `PortingGuide.md`, so that the code blocks appear indented and inside the list items

### CODE REVIEW

**Comment by mxmlnkn** [comment]
The check may have passed, but they can't check the runtime behavior of newly added function, of course,

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 1]
```diff
@@ -87,5 +87,39 @@
...
+        union ba { B b; A a; };
+        ba tmp;
+        tmp.a = x;
+        return tmp.b;
+    }
+
+} // namespace cupla
+
+#ifndef ALPAKA_ACC_GPU_CUDA_ENABLED
+#   define __int_as_float        (cupla::__A_as_B< int      , float     >)
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: Please do not align code.

Comment (mxmlnkn) [inline include/cupla/cudaToCupla/driverTypes.hpp]: Sry, I used block/vertical select/copy/paste to generate this quick and easy and it's better readable ...

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 2]
```diff
@@ -68,24 +64,58 @@
...
 // atomic functions
 #define atomicAdd(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Add>(acc, ppPointer, ppValue)
+#define atomicSub(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Sub>(acc, ppPointer, ppValue)
+#define atomicMin(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Min>(acc, ppPointer, ppValue)
+#define atomicMax(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Max>(acc, ppPointer, ppValue)
+#define atomicInc(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Inc>(acc, ppPointer, ppValue)
+#define atomicDec(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Dec>(acc, ppPointer, ppValue)
 #define atomicExch(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Exch>(acc, ppPointer, ppValue)
-
-#define uint3 ::cupla::uint3
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: please revert this line

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 3]
```diff
@@ -68,24 +64,58 @@
 
 // index renaming
 #define blockIdx                                                               \
-  static_cast<uint3>(                                                \
+  static_cast<cupla::uint3>(                                                \
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: please revert this and following lines

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 4]
```diff
@@ -68,24 +64,58 @@
...
+        ba tmp;
+        tmp.a = x;
+        return tmp.b;
+    }
+
+
+} // namespace cupla
+
+#ifndef ALPAKA_ACC_GPU_CUDA_ENABLED
+#   define __int_as_float( cupla::__A_as_B< int, float > )
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: This should be `#define __int_as_float(...) cupla::__A__as_B<int,float>(__VA_ARGS__)`

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 5]
```diff
@@ -68,24 +64,58 @@
...
+namespace cupla {
+
+
+    /* no matter how ALPAKA_FN_HOST_ACC is defined, we want to inline this */
+    template< class A, class B >
+    ALPAKA_FN_NO_INLINE_HOST_ACC inline
+    B __A_as_B( A const & x )
+    {
+        static_assert( sizeof(A) == sizeof(B), "reinterpretation assumes data types of same size!" );
+        union ba { B b; A a; };
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: why not using `reinterprete_cast<>()` insteat of the union?

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 6]
```diff
@@ -68,24 +64,58 @@
...
+ * errors when using a CUDA accelerator with alpaka.
+ * Note that there may be no spaces between the macro function name and
+ * the argument parentheses. */
+namespace cupla {
+
+
+    /* no matter how ALPAKA_FN_HOST_ACC is defined, we want to inline this */
+    template< class A, class B >
+    ALPAKA_FN_NO_INLINE_HOST_ACC inline
+    B __A_as_B( A const & x )
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: maybe `__` prefix should be removed because this is reserved for special precompiler syntax.

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 7]
```diff
@@ -68,24 +64,58 @@
...
+/* defining these as inling functions will result in multiple declaration
+ * errors when using a CUDA accelerator with alpaka.
+ * Note that there may be no spaces between the macro function name and
+ * the argument parentheses. */
+namespace cupla {
+
+
+    /* no matter how ALPAKA_FN_HOST_ACC is defined, we want to inline this */
+    template< class A, class B >
+    ALPAKA_FN_NO_INLINE_HOST_ACC inline
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: The prefix is unknown in alpaka. This is currenly defined in an unmerged branch

File: doc/PortingGuide.md [hunk 8]
```diff
@@ -61,85 +61,113 @@ Porting Step by Step
...
+   * The type of extern shared memory is always a plain pointer of the given type.
+   * In this example the type of `fooPtr` is `float*`
+   */
+  sharedMemExtern( fooPtr, float );
+
+  int bar = fooCArray2D[ threadIdx.x ][ 0 ];
+  // ...
+  ```
+
+- use `ALPAKA_FN_ACC_CUDA_ONLY` in device function definitions and add an `acc` parameter. Note that to be exact the `acc` parameter is only necessary when `alpaka` functions like `blockIdx` or `atomicMax`, ... are used.
```
Comment (psychocoderHPC) [inline doc/PortingGuide.md]: please use `ALPAKA_FN_ACC` also in the followed examples

File: doc/PortingGuide.md [hunk 9]
```diff
@@ -61,85 +61,113 @@ Porting Step by Step
...
+      template< typename T_Elem >
+      __device__ inline int deviceFunction( T_Elem x )
+      {
+          // ...
+      }
+
+  **cupla**
+
+      template< typename T_Acc, typename T_Elem >
+      ALPAKA_FN_ACC_CUDA_ONLY inline int deviceFunction( T_Acc const & acc, T_Elem x )
```
Comment (psychocoderHPC) [inline doc/PortingGuide.md]: please remove inline

File: doc/PortingGuide.md [hunk 10]
```diff
@@ -61,85 +61,113 @@ Porting Step by Step
...
+  int bar = fooCArray2D[ threadIdx.x ][ 0 ];
+  // ...
+  ```
+
+- use `ALPAKA_FN_ACC_CUDA_ONLY` in device function definitions and add an `acc` parameter. Note that to be exact the `acc` parameter is only necessary when `alpaka` functions like `blockIdx` or `atomicMax`, ... are used.
+
+  **CUDA**
+
+      template< typename T_Elem >
+      __device__ inline int deviceFunction( T_Elem x )
```
Comment (psychocoderHPC) [inline doc/PortingGuide.md]: please remove inline

File: include/cupla/cudaToCupla/driverTypes.hpp (hunk 5 above)
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: ahh I see reinterpret cast is only for pointer and reverences.

Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]:
```
return reinterprete_cast< B const & >( x );
```

Should be better and avoid the temporary copy of the variable.

Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: Please see [this post](http://stackoverflow.com/questions/16637074/c-unions-vs-reinterpret-cast)

Comment (mxmlnkn) [inline include/cupla/cudaToCupla/driverTypes.hpp]: I had this with `reinterpret_cast` before... something didn't work with it, unfortunately I didn't write down what exactly. I'll look into it.

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 11]
```diff
@@ -57,7 +53,7 @@
       ::alpaka::block::shared::st::allocVar<__VA_ARGS__, __COUNTER__>(acc)
 
 #define sharedMemExtern(ppName, ...)                                           \
-    __VA_ARGS__* ppName =                                                      \
+    __VA_ARGS__ *ppName =                                                      \
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: Could you please write ` * ` to fulfill the coding guide lines

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 12]
```diff
@@ -38,12 +38,8 @@
 
 #define dim3 cupla::dim3
 #define cudaExtent cupla::Extent
-#define cudaPos cupla::Pos
-#define cudaArray cupla::cuplaArray;
 #define cudaPitchedPtr cupla::PitchedPtr
 
-#define cudaMemcpy3DParms cupla::Memcpy3DParms
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: please revert

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 13]
```diff
@@ -38,12 +38,8 @@
 
 #define dim3 cupla::dim3
 #define cudaExtent cupla::Extent
-#define cudaPos cupla::Pos
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: please revert both lines

File: include/cupla/cudaToCupla/driverTypes.hpp (hunk 3 above)
Comment (mxmlnkn) [inline include/cupla/cudaToCupla/driverTypes.hpp]: I'm sorry, something went wrong this and with the above one. I thought I explicitly solved the merge-conflict in favor of the original.

File: doc/PortingGuide.md [hunk 14]
```diff
@@ -33,22 +33,22 @@ Porting Step by Step
...
-/* Do NOT include other headers that use CUDA runtime functions or variables
- * (see above) before this include.
- * The reason for this is that cupla renames CUDA host functions and device build in 
- * variables by using macros and macro functions.
- * Do NOT include other specific includes such as `<cuda.h>` (driver functions,
- * etc.).
- */
-#include <cuda_to_cupla.hpp>
-```
+  **CUDA include**
```
Comment (ax3l) [inline doc/PortingGuide.md]: looks like an unnecessary change. this puts the lines below in the `-` level.

File: doc/PortingGuide.md [hunk 15]
```diff
@@ -61,85 +61,113 @@ Porting Step by Step
...
-{
-    template< typename T_Acc >
-    ALPAKA_FN_ACC
-    void operator()( T_Acc const & acc, int * const ptr, float const value) const
-    {
-        // ...
-    }
-};
-```
+  **CUDA kernel**
```
Comment (ax3l) [inline doc/PortingGuide.md]: looks like an unnecessary change. this puts the lines below in the `-` level.

File: doc/PortingGuide.md [hunk 16]
```diff
@@ -61,85 +61,113 @@ Porting Step by Step
...
-
-**cupla host side kernel call**
-```C++
-// ...
-dim3 gridSize(42,1,1);
-dim3 blockSize(256,1,1);
-// extern shared memory and stream is optional
-CUPLA_KERNEL(fooKernel< 16 >)( gridSize, blockSize, 0, 0 )( ptr, 23 );
-```
+  **CUDA host side kernel call**
```
Comment (ax3l) [inline doc/PortingGuide.md]: looks like an unnecessary change. this puts the lines below in the `-` level.

File: doc/PortingGuide.md [hunk 17]
```diff
@@ -61,85 +61,113 @@ Porting Step by Step
...
- *
- * The type of extern shared memory is always a plain pointer of the given type.
- * In this example the type of `fooPtr` is `float*`
- */
-sharedMemExtern( fooPtr, float );
-
-int bar = fooCArray2D[ threadIdx.x ][ 0 ];
-// ...
-```
+  **Cuda shared memory** (in kernel)
```
Comment (ax3l) [inline doc/PortingGuide.md]: should not me indented.

**Comment by ax3l** [comment]
@mxmlnkn please provide pull-request descriptions. thanks! :sparkles:

File: doc/PortingGuide.md (hunk 16 above)
Comment (ax3l) [inline doc/PortingGuide.md]: or is this what you want?

File: doc/PortingGuide.md (hunk 15 above)
Comment (ax3l) [inline doc/PortingGuide.md]: or is this what you want?

File: doc/PortingGuide.md (hunk 17 above)
Comment (ax3l) [inline doc/PortingGuide.md]: or is this what you want?

File: doc/PortingGuide.md (hunk 15 above)
Comment (mxmlnkn) [inline doc/PortingGuide.md]: yes this is what I want. I found the document confusing to read, because the code blocks weren't indented like the list. (in rendered form) It seemed to me like the list ended and I wondered what all these code blocks meant outside of the list and so on ...

**Comment by mxmlnkn** [comment]
Well. I edited the description, but it's basically the same as the commit comments, so I didn't bother before.

File: doc/PortingGuide.md (hunk 15 above)
Comment (ax3l) [inline doc/PortingGuide.md]: makes sense!

**Comment by ax3l** [comment]
thanks, descriptions are for humans that
- review (now, as an introduction and overview/additional ressource)
- write a change log (much later, there is not much time at this point)

to get an understanding of what changed without reading all the commit messages again.

**Comment by ax3l** [comment]
@mxmlnkn I updated your PR description to help you a bit

**Comment by ax3l** [comment]
again, thank you for the PR! :+1: some comments from both of us are still inline.

File: include/cupla/cudaToCupla/driverTypes.hpp (hunk 6 above)
Comment (mxmlnkn) [inline include/cupla/cudaToCupla/driverTypes.hpp]: Hm, but CUDA uses it too, so if it was, then it would break something anyway.

File: include/cupla/cudaToCupla/driverTypes.hpp (hunk 5 above)
Comment (mxmlnkn) [inline include/cupla/cudaToCupla/driverTypes.hpp]: A prior implementation looked like this:

    #define __int_as_float(x) reinterpret_cast< float const & >( x )
    old = atomicCAS( (int*) rdpTarget, assumed,
                __float_as_int( f( __int_as_float(assumed), rValue ) )

But this gets me several compiler warnings:

    imresh/src/imresh/algorithms/cuda/cudaVectorReduce.tpp:88:68: warning: dereferencing type-punned pointer will break strict-aliasing rules [-Wstrict-aliasing]
                     __float_as_int( f( __int_as_float(assumed), rValue ) )
                                                                        ^

Comment (mxmlnkn) [inline include/cupla/cudaToCupla/driverTypes.hpp]: The error doesn't seem to appear, if the `reinterpret_cast` is encapsulated inside a function, though, so I guess there are no arguments against it.

File: include/cupla/cudaToCupla/driverTypes.hpp (hunk 4 above)
Comment (mxmlnkn) [inline include/cupla/cudaToCupla/driverTypes.hpp]: well, I just applied it already, but why variational arguments? `A_AS_B` allows only 1 argument anyway. Furthermore the above define should find and replace all `__int_as_float` with `( cupla::__A_as_B< int, float > )` if I understand correctly, so it works exactly the same, because then instances of :

    __int_as_float(x);

will be converted to:

    ( cupla::__A_as_B< int, float > )(x)

although this is much much lower level.

**Comment by mxmlnkn** [comment]
All comments should be applied and integrated now.

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 18]
```diff
@@ -86,6 +86,38 @@
...
+
+// recast functions
+/* defining these as inling functions will result in multiple declaration
+ * errors when using a CUDA accelerator with alpaka.
+ * Note that there may be no spaces between the macro function name and
+ * the argument parentheses. */
+namespace cupla {
+
+
+    template< class A, class B >
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: please use `typename` instead of the `class` keyword

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 19]
```diff
@@ -86,6 +86,38 @@
...
+#define atomicDec(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Dec>(acc, ppPointer, ppValue)
 #define atomicExch(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Exch>(acc, ppPointer, ppValue)
+#define atomicCAS(ppPointer,ppCompare,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Cas>(acc, ppPointer, ppCompare, ppValue)
 
 #define uint3 ::cupla::uint3
+
+// recast functions
+/* defining these as inling functions will result in multiple declaration
+ * errors when using a CUDA accelerator with alpaka.
+ * Note that there may be no spaces between the macro function name and
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: This comment looks like it it placed wrong here. There is no macro function defined here.

**Comment by psychocoderHPC** [comment]
Two small things than we can merge it.

File: include/cupla/cudaToCupla/driverTypes.hpp [hunk 20]
```diff
@@ -86,6 +86,38 @@
...
+#define atomicMax(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Max>(acc, ppPointer, ppValue)
+#define atomicInc(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Inc>(acc, ppPointer, ppValue)
+#define atomicDec(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Dec>(acc, ppPointer, ppValue)
 #define atomicExch(ppPointer,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Exch>(acc, ppPointer, ppValue)
+#define atomicCAS(ppPointer,ppCompare,ppValue) ::alpaka::atomic::atomicOp<::alpaka::atomic::op::Cas>(acc, ppPointer, ppCompare, ppValue)
 
 #define uint3 ::cupla::uint3
+
+// recast functions
+/* defining these as inling functions will result in multiple declaration
```
Comment (psychocoderHPC) [inline include/cupla/cudaToCupla/driverTypes.hpp]: If there are problems with `inline` than we can also write this as a class with a static method.

**Comment by mxmlnkn** [comment]
I hope this is all now :)

**Comment by psychocoderHPC** [comment]
big thanks!
